src/firmware/app/control/python_controller/difference_equation.py

```diff
@@ -0,0 +1,162 @@
+        ]
+
+        # The first element in the output coefficient list corresponds to the system output for a given input,
+        # so we will remove it here so it has no contribution
+        del self.__output_coefficients[0]
+
+    def calculate_output(
+        self, output_history: List[float], input_history: List[float]
+    ) -> float:
+        """ 
+        Calculates the output of the difference equation given the list of previous(and current) outputs
+        and inputs to the system
+        
+        :param output_history The last M outputs, where M is the order of the original TF's denominator
+        aka self.__denominator_order
+        
+        :param input_history The last N inputs, where N is the order of the original TF's numerator
+        aka self.__numerator_order
+        """
+
+        # Calcualte the effect of the input history has on the output
+        input_response = 0
+        history_index_shift = self.__denominator_order - self.__numerator_order
+        for i in range(history_index_shift, self.__denominator_order):
+            input_response += (
+                input_history[i] * self.__input_coefficients[i - history_index_shift]
+            )
+
+        # Calculate the effect of the output history has on the output
+        output_response = [
+            coefficient * input_value
+            for coefficient, input_value in zip(
+                self.__output_coefficients, output_history
+            )
+        ]
+
+        return input_response - functools.reduce(lambda a, b: a + b, output_response)
+
+    def get_input_order(self):
+        """
+        Returns the order of the input polynomial
+        """
+        return self.__numerator_order
+
+    def get_output_order(self):
+        """
+        Returns the order of the output polynomial
+        """
+        return self.__denominator_order
+
+    def get_input_coefficients(self):
+        """
+        Returns the Coefficients of the input polynomial from the least negative Z 
+        exponent to the most negative. (Z^0, Z^-1, Z^-2...)
+        """
+        return self.__input_coefficients
+
+    def get_output_coefficients(self):
+        """
+        Returns the Coefficients of the output polynomial from the least negative Z 
+        exponent to the most negative. (Z^0, Z^-1, Z^-2...)
+        """
+        return self.__output_coefficients
+
+
+class DifferenceEquationSimulator:
+    def __init__(self, difference_equation: DifferenceEquation):
+        """ 
+        Creates a difference equation simulator based on the input difference equation transfer function
+
+        :param difference_equation The difference equation transfer function (Z domain) representing the system
+        """
+
+        # Create the circular buffers that are sized to contain all the needed information
+        # required to run the difference equation
+        self.__input_buffer = collections.deque(
+            maxlen=difference_equation.get_output_order()
```

Comment:
Not in this case - I should add a comment.

I should remove the input order almost completely. The input buffer should be the length of the denominator order in case the numerator is not the same order as the denominator